File: app/src/main/java/sugtao4423/yandereviewer/EndlessScrollListener.java
package sugtao4423.yandereviewer;

import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;

public abstract class EndlessScrollListener extends RecyclerView.OnScrollListener{

    int visibleThreshold = 5;
    int visibleItemCount, totalItemCount;
    private int previousTotal = 0;
    private boolean loading = true;
    private boolean stopOnLoadMore = false;
    private int current_page = 0;

    private GridLayoutManager mGridLayoutManager;

    public EndlessScrollListener(GridLayoutManager gridLayoutManager){
        this.mGridLayoutManager = gridLayoutManager;
    }

    public void setVisibleThreshold(int visibleThreshold){
        this.visibleThreshold = visibleThreshold;
    }

    public void setStopOnLoadMore(boolean isStop){
        this.stopOnLoadMore = isStop;
    }

    @Override
    public void onScrolled(RecyclerView recyclerView, int dx, int dy){
        super.onScrolled(recyclerView, dx, dy);

        if(stopOnLoadMore)
            return;

        visibleItemCount = recyclerView.getChildCount();
        totalItemCount = mGridLayoutManager.getItemCount();
        int lastVisibleItem = mGridLayoutManager.findLastVisibleItemPosition();

        if(loading && totalItemCount > previousTotal){
            loading = false;
            previousTotal = totalItemCount;
        }

        if(!loading && (lastVisibleItem + visibleThreshold) > totalItemCount){
            current_page++;
            onLoadMore(current_page);
            loading = true;
        }
    }

    public void resetState(){
        this.current_page = 0;
        this.previousTotal = 0;
        this.loading = true;
        this.stopOnLoadMore = false;
    }

    public abstract void onLoadMore(int current_page);
}

File: app/src/main/java/sugtao4423/yandereviewer/EndlessScrollListener.kt
package sugtao4423.yandereviewer

import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView

abstract class EndlessScrollListener(private val gridLayoutManager: GridLayoutManager) : RecyclerView.OnScrollListener() {

    var visibleThreshold = 5
    var visibleItemCount = -1
    var totalItemCount = -1
    private var previousTotal = 0
    private var loading = true
    var stopOnLoadMore = false
    private var currentPage = 0

    override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
        super.onScrolled(recyclerView, dx, dy)

        if (stopOnLoadMore) {
            return
        }

        visibleItemCount = recyclerView.childCount
        totalItemCount = gridLayoutManager.itemCount
        val lastVisibleItem = gridLayoutManager.findLastVisibleItemPosition()

        if (loading && totalItemCount > previousTotal) {
            loading = false
            previousTotal = totalItemCount
        }

        if (!loading && (lastVisibleItem + visibleThreshold) > totalItemCount) {
            currentPage++
            onLoadMore(currentPage)
            loading = true
        }
    }

    fun resetState() {
        this.currentPage = 0
        this.previousTotal = 0
        this.loading = true
        this.stopOnLoadMore = false
    }

    abstract fun onLoadMore(currentPage: Int)

}